File: python/langchain-payment-agent/utils/mandate_storage.py
"""
Simple mandate storage - stores mandate tokens to reuse across runs
"""
import json
import os
from datetime import datetime
from pathlib import Path

STORAGE_FILE = Path(__file__).parent.parent / ".agentgatepay_mandates.json"

def save_mandate(agent_id: str, mandate_data: dict):
    """Save mandate for reuse"""
    storage = _load_storage()
    storage[agent_id] = {
        'mandate_token': mandate_data.get('mandate_token'),
        'expires_at': mandate_data.get('expires_at'),
        'budget_remaining': mandate_data.get('budget_remaining') or mandate_data.get('budget_usd'),
        'budget_usd': mandate_data.get('budget_usd'),
        'saved_at': datetime.now().isoformat()
    }
    _save_storage(storage)

def get_mandate(agent_id: str) -> dict:
    """Get stored mandate if valid"""
    storage = _load_storage()
    if agent_id not in storage:
        return None

    mandate = storage[agent_id]
    expires_at = mandate.get('expires_at', 0)

    if datetime.now().timestamp() > expires_at:
        del storage[agent_id]
        _save_storage(storage)
        return None

    return mandate

def clear_mandate(agent_id: str):
    """Clear stored mandate"""
    storage = _load_storage()
    if agent_id in storage:
        del storage[agent_id]
        _save_storage(storage)

def _load_storage() -> dict:
    if not STORAGE_FILE.exists():
        return {}
    try:
        return json.loads(STORAGE_FILE.read_text())
    except:
        return {}

def _save_storage(data: dict):
    STORAGE_FILE.write_text(json.dumps(data, indent=2))

File: javascript/langchain-payment-agent/utils/mandateStorage.ts
/**
 * Mandate Storage Utilities
 *
 * Persistent storage for AP2 mandates across agent runs.
 * Uses JSON file similar to Python implementation.
 */

import { readFileSync, writeFileSync, existsSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Storage file path (in parent directory)
const MANDATE_STORAGE_FILE = join(__dirname, '..', '.agentgatepay_mandates.json');

export interface StoredMandate {
  mandate_token: string;
  budget_usd: number;
  budget_remaining: number;
  purpose?: string;
  created_at?: string;
  [key: string]: any;
}

interface MandateStorage {
  [agentId: string]: StoredMandate;
}

/**
 * Load all mandates from storage file
 */
function loadMandates(): MandateStorage {
  if (!existsSync(MANDATE_STORAGE_FILE)) {
    return {};
  }

  try {
    const data = readFileSync(MANDATE_STORAGE_FILE, 'utf8');
    return JSON.parse(data);
  } catch (error) {
    console.error('⚠️  Failed to load mandates:', error);
    return {};
  }
}

/**
 * Save all mandates to storage file
 */
function saveMandates(mandates: MandateStorage): void {
  try {
    writeFileSync(MANDATE_STORAGE_FILE, JSON.stringify(mandates, null, 2));
  } catch (error) {
    console.error('⚠️  Failed to save mandates:', error);
  }
}

/**
 * Save a mandate for an agent
 */
export function saveMandate(agentId: string, mandate: StoredMandate): void {
  const mandates = loadMandates();
  mandates[agentId] = {
    ...mandate,
    created_at: mandate.created_at || new Date().toISOString()
  };
  saveMandates(mandates);
}

/**
 * Get a mandate for an agent
 */
export function getMandate(agentId: string): StoredMandate | null {
  const mandates = loadMandates();
  return mandates[agentId] || null;
}

/**
 * Clear a mandate for an agent
 */
export function clearMandate(agentId: string): void {
  const mandates = loadMandates();
  delete mandates[agentId];
  saveMandates(mandates);
}

/**
 * Clear all mandates
 */
export function clearAllMandates(): void {
  saveMandates({});
}
